pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+	// m is a map of all the currently traced entries for this replica. The
+	// aggregate size of the map is equal to or less than numRegistered. We add
+	// to numRegistered before we update m, and delete from m before we remove
+	// from numRegistered to keep this invariant.
+	m syncutil.Map[kvpb.RaftIndex, traceValue]
+	// numRegistered is the number of currently registered traces for this
+	// store, not this replica. The number of registered will always be less
+	// than the MaxConcurrentRaftTraces setting. If the setting is lowered, we
+	// flush all traces on all replicas.
+	numRegistered *atomic.Int64
+	// This is the ambient context for the replica and is used for remote
+	// traces. It contains the replica/range information. On each trace we
+	// additionally append the unique trace/span IDs.
+	ctx context.Context
+	st  *cluster.Settings
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context, st *cluster.Settings, numRegistered *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, st: st, numRegistered: numRegistered}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages. 10 is a
+// reasonable default that balances usefulness with performance impact. It isn't
+// expected that this limit will normally be hit.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// maybeRegister checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The invariant is
+// that numRegistered <= numAllowed.  This method will return true if we can
+// keep the invariant and added one to the number registered, otherwise it will
+// return false.
+func (r *RaftTracer) maybeRegister() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegistered := r.numRegistered.Load()
+
+	// The maximum number of traces has been reached. We don't register this
+	// trace and return false.
+	if numRegistered == numAllowed {
+		return false
+	}
+
+	// This can only happen if numAllowed has changed. If this happens flush all
+	// our current traces and don't register this request.
+	for numRegistered > numAllowed {
+		r.FlushAll()
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	return r.numRegistered.CompareAndSwap(numRegistered, numRegistered+1)
```

Comment:
I was OK with missing a registration in a rare case of a race (since it can only happen if there are multiple in-progress attempts and at least one will win). I wasn't too worried about the cost here, as if we are already here we almost always will succeed on this CAS and are going to enable expensive tracing. The checks above will remove the common cases. 

I understand the concern about the global variable, but I'm not sure of a good alternative for the 100K range case. Even having 1 trace per range will be too many in that case, and for a KV like workload it is likely to be in that situation. Maybe we can chat at the standup today.